ITR/data/base_providers.py

```diff
@@ -516,46 +569,53 @@ def _compute_missing_historic_ei(self, companies, historic_data):
             if scope_projections['S1'] and scope_projections['S2'] and not scope_projections['S1S2']:
                 results = scope_dfs['S1'] + scope_dfs['S2']
                 units = f"{results.values[0].u:~P}"
                 projections = [IProjection(year=year, value=value) for year, value in results.items()
                                if year in range(self.projection_controls.BASE_YEAR, self.projection_controls.TARGET_YEAR+1)]
-                scope_projections['S1S2'] = ICompanyEIProjections(ei_metric={'units': units}, projections=projections)
+                scope_projections['S1S2'] = ICompanyEIProjections(ei_metric=units, projections=projections)
             # FIXME: do we really need to do this?  We're going to migrate S3 to S1S2 and ignore S1S2S3...
             if scope_projections['S1S2'] and scope_projections['S3'] and not scope_projections['S1S2S3']:
                 results = scope_dfs['S1S2'] + scope_dfs['S3']
                 units = f"{results.values[0].u:~P}"
                 projections = [IProjection(year=year, value=value) for year, value in results.items()
                                if year in range(self.projection_controls.BASE_YEAR, self.projection_controls.TARGET_YEAR+1)]
-                scope_projections['S1S2S3'] = ICompanyEIProjections(ei_metric={'units': units}, projections=projections)
+                scope_projections['S1S2S3'] = ICompanyEIProjections(ei_metric=units, projections=projections)
             company.projected_intensities = ICompanyEIProjectionsScopes(**scope_projections)
 
     def _standardize(self, intensities: pd.DataFrame) -> pd.DataFrame:
         # When columns are years and rows are all different intensity types, we cannot winsorize
         # Transpose the dataframe, winsorize the columns (which are all coherent because they belong to a single variable/company), then transpose again
-        intensities = intensities.T
-        for col in intensities.columns:
-            s = intensities[col]
-            ei_units = f"{s.loc[s.first_valid_index()].u:~P}"
-            if s.notnull().any():
-                with warnings.catch_warnings():
-                    warnings.simplefilter("ignore")
-                    intensities[col] = s.map(lambda x: Q_(np.nan, ei_units)
-                    if x.m is np.nan or x.m is pd.NA else x).astype(f"pint[{ei_units}]")
+        na_intensities = intensities.apply(lambda x: x.isna().all(), axis=1)
+        intensities = intensities[~na_intensities].T
+        if ITR.HAS_UNCERTAINTIES:
+            for col in intensities.columns:
+                pa = PA_._from_sequence(intensities[col])
+                pa_units = pa[0].u
+                pa_nans = ITR.isnan(pa.data)
+                if pa_nans.any():
+                    if [isinstance(pt.m, ITR.UFloat) for pt in pa[np.where(pa_nans)]]:
+                        u_pa_data = [ITR._ufloat_nan if pn else pt if isinstance(pt, ITR.UFloat) else ITR.ufloat(pt, 0)
```

Comment:
why do we need separate `ITR.UFloat` and `ITR.ufloat`?
if these classes are different, shouldn't they have different names?

Reply:
I've explained this before.  UFloat is a type, ufloat is a constructor.  ufloat(1,0) produces a UFloat object with nominal value 1 and uncertainty of 0.

Reply:
Why can't this constructor be a part of s class? Having stand-alone functions in object oriented python module may lead to more confusions in future

Reply:
I didn't write the Uncertainties package, I'm only using it as it is documented.  There's a lot of magic that happens when using wrappers (or ExtensionArrays).  I don't want to break/reimplement that magic by creating some new set of interfaces that, themselves, would need to be documented.

Reply:
i see. My mistake.
it's reference as `ITR.UFloat` and `ITR.ufloat`, made me think, that it's written by us.
why don't we refer to them as `uncertainties.UFloat` and `uncertainties.ufloat`, then?
if we find a way to *not* `import ITR` directly, will we be able to switch to `UFloat` and `ufloat`, in a way, which will make clear, that they are not written by us? like

~~~~
from uncertainties import ufloat, UFloat
~~~~

?

Reply:
Because if uncertainties is not installed, it will fail.  The idea was to let ITR handle the polymorphism of floats vs. uncertainties, creating an abstraction layer that implements everything consistently so the user doesn't have to special-case their own code every time they touch a number.

Reply:
In a different review comment, i suggested a dedicated package ‘ITR/uncert‘, which should fix the polymorphism question, w/o a need to import the whole ’ITR’ package within ITR package.

WRT this ufloat use-case, the call will become ’uncert.ufloat()’.
This will solve my questions to ’ITR.float()’.
Will it it also solve your needs?

Reply:
i figured out from pep-8, that your approach is fine within python guidelines.
i don't have questions to `import ITR` any more